src/client/connection.rs

```diff
@@ -295,32 +329,36 @@ impl Connection {
     /// channel) and creates a stream of packets to send on network
     fn network_reply_stream(&self, network_stream: SplitStream<MqttFramed>) -> impl PacketStream {
         let mqtt_state = self.mqtt_state.clone();
+        let mqtt_state_notification = self.mqtt_state.clone();
         let keep_alive = self.mqttoptions.keep_alive();
         let notification_tx = self.notification_tx.clone();
-
         let network_stream = network_stream
             .map_err(NetworkError::Io)
             .and_then(move |packet| {
                 debug!("Incoming packet = {:?}", packet_info(&packet));
-                let reply = mqtt_state.borrow_mut().handle_incoming_mqtt_packet(packet);
-                future::result(reply)
+                let mut mqtt_state = mqtt_state.borrow_mut();
+                mqtt_state.handle_incoming_mqtt_packet(packet)
             })
             .and_then(move |(notification, reply)| {
-                handle_notification(notification, &notification_tx);
-                future::ok(reply)
+                let mqtt_state = mqtt_state_notification.borrow();
+                match notification {
+                    Notification::None => (),
+                    // Ignore error on notification_tx send, since the receiver can be dropped at any time
+                    _ if mqtt_state.send_notifications() => drop(notification_tx.send(notification)),
```

Comment:
This looks like this will block the event loop. What happens when the receiver is not dropped and the channel is full?

Reply:
Yes. This blocks the loop. As far a I got it that's ok in order to backpressure on the broker.
A option would be to use a unbounded channel - but this is probably not what we want here.

I'm not fluent in the MQTT spec: What is a broker allowed to do with clients that back pressure?

Reply:
Blocking the event loop is not ok just because the receiver is not able to catch up. Pings should still continue to happen to prevent broker disconnecting the client and publisher should not be slow because the receiver is doing some heavy computation.

This is getting a little tricky than I anticipated. I'll give proper thought to this and ping you in a few days.

Reply:
Hm. You're right. I forgot the pings...sorry. Will think about that too.

Reply:
Cool :). Sorry for the back and forths on this

Reply:
Options regarding the pings:
1. Accept that an overload of the client will stop sending pings. The broker will probably disconnect in such a situation (or has to according the spec?). The timeouts are quite relaxed to be fine with a short delayed ping. This is the behaviour on `notification_refactor` (this PR).
2. A `unbound` notification queue: sorry - bad idea... 
3. Discard notifications when the notification queue is full. Behaviour on `master`. I'd really prefer option 1 one over 3 since at least you see what happened.
4. Another idea is to split the notification queue into several ones to give the client more flexibility what to handle or which of the `Receivers` to drop. e.g `Notification::Puback` could be less interesting that `Notification::Connected` or `Notification::Disconnected(_)`. Here still the decision whether a `Notification` can be discarded or not has to be done.
5. Configuration options to choose between option 1 and 3. Should not be a `feature` but could be added to the config struct.

What do you think?

Reply:
Hi @flxo. Sorry for the delay on this. Coming to the options that you've mentioned

1. Blocking the event loop, in general, isn't a good idea.  It's not just about pings and disconnects. If someone is doing publishes from a different thread and the receiver is doing heavy computations, publishers won't progress. Just because crossbeam channel is full, blocking the publisher doesn't make any sense. Also, the eventloop block will keep interfering with any future features we might need on the eventloop.

2. Agreed :)

4. Sounds good

Personally, I feel that we should provide an option for `future channels` along with `crossbeam channel` for notifications. Future channels won't have this problem of blocking the eventloop while doing a blocking send. Also we won't lose data silently. If people choose to use crossbeam channel, the get the flexibility of select but might miss incoming data.

I'm open to discussion on this. May on a new issue :)

Also, I'll busy this month and won't be able to do a lot of rumqtt work. But we can try to reach a conclusion this month and implement it in the next